### PR TITLE
Pursuit refactor

### LINKED ISSUE #4492: Pursuit issues
### Description

Interaction that is likely broken (unconfirmed):

> Pursuit will now hit any adjacent opponent that attempts to switch out (but not more than one per turn), regardless of who it originally targeted. This effect bypasses redirection effects such as [Follow Me](https://bulbapedia.bulbagarden.net/wiki/Follow_Me_(move)).

Missing effect:
> If the target is using [Volt Switch](https://bulbapedia.bulbagarden.net/wiki/Volt_Switch_(move)) or [Parting Shot](https://bulbapedia.bulbagarden.net/wiki/Parting_Shot_(move)), Pursuit's special effect will activate only if the user of Pursuit would usually act later than the target (its power will only be doubled in that case).
Pursuit is now affected by King's Rock.

### Version

upcoming (Edge)

### Upcoming/master Version

_No response_

### Discord contact info

_No response_

### LINKED ISSUE #5680: Pursuit does not trigger Life Orb damage if hitting a switch out
### Description

The Pursuit user is supposed to still take Life Orb damage when used on a switching opponent.

https://github.com/user-attachments/assets/925eb642-5d87-4a63-9465-facd3c7a5560

From Run&Bun discord

### Version

master (default, unreleased bugfixes)

### Upcoming/master Version

dc740b4e9a

### Discord contact info

hedara

### PR DESCRIPTION
**Pull request by PhallenTree**
Refactored Pursuit to have a similar implementation to After You/Pledge.

In short, this new implementation:

1. Interrupts the switch if Pursuit can happen.
2. Makes the next battler (according to turn order) to move with Pursuit to be the next to act.
3. Once a battler has used Pursuit in this state, `MOVEEND_PURSUIT_NEXT_ACTION` will decide the next action:
- If there is another Pursuiter and the battler who was trying to switch is still alive, repeat from step 2.
- If there isn't another Pursuiter, finish the switch.
- If the battler who tried to switch is no longer alive, the turn progresses as normal, with the replacement switching in at the end of the turn, as usually happens in modern generations.

https://github.com/user-attachments/assets/0074be4e-36c7-4362-9361-6e727a1b736d

Adds missing effect:

> If the target is using Volt Switch/U-Turn or Parting Shot, Pursuit's special effect will activate only if the user of Pursuit would usually act later than the target (its power will only be doubled in that case).

Added to `gBattleStruct`:
- `pursuitTarget` - a bitfield that stores the target and is also the way used to check if Pursuit should double in power, ignore accuracy checks, etc.;
- `pursuitSwitchByMove` - basically a boolean that stores whether the switch was caused by a move or if it was the actual Switching action;
- `pursuitStoredSwitch` - stores the mon that would come in if Pursuit didn't delay it.

Added to `moveend`:
- `MOVEEND_PURSUIT_NEXT_ACTION` (after `MOVEEND_CLEAR_BITS`) - if the switching target is alive, change turn order to make the next fastest Pursuiter act next if possible, once there are no more Pursuiters to act and if the Pursuit target is alive, do the switch out. If the Pursuit target isn't alive, the next Pursuiter does not get their order of turn changed and the next Pokemon does not come in until the end of the turn.

Fixed outdated switching mechanic in Pursuit's current implementation that caused the replacing mon to switch in after the switching mon fainted.

Added config:
- `B_PURSUIT_TARGET` for Gen 4+ effect:

> Pursuit will now hit any adjacent opponent that attempts to switch out (but not more than one per turn), regardless of who it originally targeted.

Other notes:
There is no indication that Chilly Reception would be affected by Pursuit, so I used a separate copy of `BattleScript_MoveSwitch` for it.
There is no indication that Flip Turn would be affected by Pursuit, but other moves with the same effect are affected (Volt Switch/U-Turn), therefore I assumed it would be affected by it.
`Cmd_decrementmultihit` became unused.

## Issue(s) that this PR fixes
Fixes #5680 , fixes #4492 

## Feature(s) this PR does NOT handle:
- Volt Switch and U-Turn cause the battler to be invisible during Pursuit, due to their animations.
- Attempted Switch -> Mega Evolution before Pursuit may be incorrect, this implementation tries to do all battlers' gimmicks before any moves occur, so all battlers Mega Evolutions occur before the first Pursuit which may be incorrect.

## **Discord contact info**
PhallenTree

### CODE REVIEW

**Comment by AlexOn1ine** [comment]
Great work!

**Review by hedara90 — COMMENTED** [review]
The only questions I have is with the Gimmick case and Teleport/Chilly Reception.
For the gimmick case, I don't know if other battlers using a gimmick should happen at the same time as the Pursuit user.
For Teleport/Chilly Reception, it's a question if it's more like Baton Pass or more like Parting Shot since the modern Teleport behavior doesn't exist at the same time as Pursuit.
Arguments can be made for both cases, so it's not actually blocking.

Other than that it looks very good, I just added some more test suggestions for checking pursuit behavior.

File: test/battle/move_effect/pursuit.c
Comment (PhallenTree) [reply test/battle/move_effect/pursuit.c:544]: It doesn't handle the increased power but does handle the increased priority. Should this be removed?

**Comment by PhallenTree** [comment]
> For the gimmick case, I don't know if other battlers using a gimmick should happen at the same time as the Pursuit user.

It makes a bit more sense to me that they shouldn't so I will fix that.

> For Teleport/Chilly Reception, it's a question if it's more like Baton Pass or more like Parting Shot since the modern Teleport behavior doesn't exist at the same time as Pursuit. Arguments can be made for both cases, so it's not actually blocking.

According to Bulbapedia, Teleport should not change Pursuit's priority, also I believe both are present with similar to current behavior in Let's Go Pikachu/Eevee, which I assumed is where Bulbapedia got this information from.
As for Chilly Reception, I could make a config for whether or not it should be affected by Pursuit, though I believe maintainers should decide whether it should be affected, shouldn't be affected or should be left to developers as a config.

File: test/battle/move_effect/pursuit.c
Comment (hedara90) [reply test/battle/move_effect/pursuit.c:613]: Ah, makes sense, it can remain then.

**Comment by hedara90** [comment]
> According to Bulbapedia, Teleport should not change Pursuit's priority, also I believe both are present with a similar behavior to the current one in Let's Go Pikachu/Eevee, which I assumed is where Bulbapedia got this information from.

I can't find Pursuit in any move lists for LGPE. So I'd say it's inconclusive.
> I could make a config for Chilly Reception if it should or shouldn't be affected by Pursuit, though I believe maintainers should decide whether it should be affected, shouldn't be affected or should be left to developers as a config.

It seems like a simple enough change to make Chilly Reception/Teleport trigger Pursuit in the rare case it could happen. It can be discussed at another time since it's an ambigous interaction.

**Review by hedara90 — APPROVED** [review]
Looks great.
Untestable interactions cans be changed at a later date after discussion, since they're ambigous.

**Pull request merged**
